### PR TITLE
Add pascal matrix to jax.scipy.linalg

### PR DESCRIPTION
**Pull request by mattbahr**
This PR adds an implementation of the Pascal matrices to `jax.scipy.linalg`. The output approximates the binomial coefficients, but for higher values of `n` (greater than about 10), the output diverges more dramatically from the `scipy` output. I thought this might be ok as it's documented [here](https://docs.jax.dev/en/latest/_autosummary/jax.scipy.special.factorial.html) that JAX does not support exact factorials, but I'm open to other ideas if we want to get the output closer to the exact coefficients.

Addresses part of the request in #10144

### CODE REVIEW

File: jax/_src/scipy/linalg.py [hunk 1]
```diff
@@ -2182,3 +2182,67 @@ def hilbert(n: int) -> Array:
...
+    return L_n.T
+
+  return jnp.dot(L_n, L_n.T)
+
+@jit
+def _binom(n, k):
+  a = _factorial(n)
+  b = _factorial(n - k)
+  c = _factorial(k)
+  return a / (b * c)
```
Comment (jakevdp) [inline jax/_src/scipy/linalg.py:2244]: Numerically, I think it would be better to keep this computation in log space, evaluating a single exponential at the end.

File: tests/linalg_test.py [hunk 2]
```diff
@@ -2329,6 +2329,21 @@ def testSymmetricProduct(self, shape, dtype, symmetrize_output):
     self.assertAllClose(
         new_product_with_batching, old_product, atol=atol)
 
+  @jtu.sample_product(
+    n=[0, 1, 5, 10],
+    kind=["symmetric", "lower", "upper"],
+  )
+  def testPascal(self, n, kind):
+    with jax.numpy_dtype_promotion("standard"):
```
Comment (jakevdp) [inline tests/linalg_test.py:2337]: This function should not do any implicit dtype promotion, so we should remove this `with` statement & fix the implementation if that causes failures.

**Review by jakevdp — CHANGES_REQUESTED** [review]
Thanks for the contribution! A few changes below and then we can get this merged 🚀

File: jax/_src/scipy/linalg.py (hunk 1 above)
Comment (mattbahr) [reply jax/_src/scipy/linalg.py]: Good idea! Much simpler and more stable

File: tests/linalg_test.py (hunk 2 above)
Comment (mattbahr) [reply tests/linalg_test.py]: My mistake, didn't think to set the `dtype` on the initial range. Thanks for the review!

File: jax/_src/scipy/linalg.py [hunk 3]
```diff
@@ -2182,3 +2182,63 @@ def hilbert(n: int) -> Array:
...
+  Args:
+    n: the size of the matrix to create.
+    kind: (optional) must be one of ``lower``, ``upper``, or ``symmetric`` (default).
+
+  Returns:
+    A Pascal matrix of shape ``(n, n)``
+
+  Examples:
+    >>> jax.scipy.linalg.pascal(5)
+    Array([[0.99999905, 0.99999905, 0.99999905, 0.99999905, 0.99999905],
```
Comment (jakevdp) [inline jax/_src/scipy/linalg.py:2204]: Outputs like this can be problematic in docs, because rounding may differ on different platforms causing doctests to fail. You can address this by locally setting `printoptions`; you can find examples elsewhere in this file e.g.: https://github.com/jax-ml/jax/blob/a7d46e6acc4aecee92dcfe68a8d3d86d21b3db3c/jax/_src/scipy/linalg.py#L805-L813

Comment (mattbahr) [reply jax/_src/scipy/linalg.py]: Updated with the precision option set. Thanks!

File: jax/_src/scipy/linalg.py [hunk 4]
```diff
@@ -2182,3 +2182,67 @@ def hilbert(n: int) -> Array:
...
+
+  Returns:
+    A Pascal matrix of shape ``(n, n)``
+
+  Examples:
+    >>> L = jax.scipy.linalg.pascal(3, kind="lower")
+    >>> U = jax.scipy.linalg.pascal(4, kind="upper")
+    >>> P = jax.scipy.linalg.pascal(5)
+    >>> with jnp.printoptions(precision=3):
+    ...   print(L)
```
Comment (jakevdp) [inline jax/_src/scipy/linalg.py:2207]: I don't think `L`, `U`, and `P` make sense as variable names here (the example I pointed to was LU-decomposition, where `L` and `U` do make sense). I think it would be clearer to print these directly, e.g.
```python
>>> with jnp.printoptions(precision=3):
...   print(jax.scipy.linalg.pascal(3, kind="lower"))
```

Comment (mattbahr) [reply jax/_src/scipy/linalg.py]: I was thinking Lower, Upper, and Pascal, but makes sense that it could be confusing with the same variable names used elsewhere. This works for me :)

**Review by jakevdp — APPROVED** [review]
Looks great, thanks!

**Comment by jakevdp** [comment]
We're seeing failures on certain GPU architectures, like this:
```
LaxLinalgTest.testPascal10 (n=10, kind='symmetric')

<...>

Mismatched elements: 41 / 100 (41%)
Max absolute difference among violations: 57.265625
Max relative difference among violations: 0.00165649
 ACTUAL: array([[1.000000e+00, 1.000000e+00, 1.000000e+00, 1.000000e+00,
        1.000000e+00, 1.000000e+00, 1.000000e+00, 1.000000e+00,
        1.000000e+00, 1.000000e+00],...
 DESIRED: array([[1.000e+00, 1.000e+00, 1.000e+00, 1.000e+00, 1.000e+00, 1.000e+00,
        1.000e+00, 1.000e+00, 1.000e+00, 1.000e+00],
       [1.000e+00, 2.000e+00, 3.000e+00, 4.000e+00, 5.000e+00, 6.000e+00,...
```
I suspect this comes from the fact that on GPU, the default matmul precision is lower than float32, and so the `"centered"` version can have lower precision.

The best fix would probably be to decorate the test with `@jax.default_matmul_precision("float32")` – can you give that a try?

**Comment by mattbahr** [comment]
> We're seeing failures on certain GPU architectures, like this:
> 
> ```
> LaxLinalgTest.testPascal10 (n=10, kind='symmetric')
> 
> <...>
> 
> Mismatched elements: 41 / 100 (41%)
> Max absolute difference among violations: 57.265625
> Max relative difference among violations: 0.00165649
>  ACTUAL: array([[1.000000e+00, 1.000000e+00, 1.000000e+00, 1.000000e+00,
>         1.000000e+00, 1.000000e+00, 1.000000e+00, 1.000000e+00,
>         1.000000e+00, 1.000000e+00],...
>  DESIRED: array([[1.000e+00, 1.000e+00, 1.000e+00, 1.000e+00, 1.000e+00, 1.000e+00,
>         1.000e+00, 1.000e+00, 1.000e+00, 1.000e+00],
>        [1.000e+00, 2.000e+00, 3.000e+00, 4.000e+00, 5.000e+00, 6.000e+00,...
> ```
> 
> I suspect this comes from the fact that on GPU, the default matmul precision is lower than float32, and so the `"centered"` version can have lower precision.
> 
> The best fix would probably be to decorate the test with `@jax.default_matmul_precision("float32")` – can you give that a try?

Sure thing, just updated! I also decreased the tolerance for the test to 1e-3. That's what I originally intended it to be. I must have accidentally increased it at some point.

**Review by jakevdp — APPROVED** [review]

File: tests/linalg_test.py [hunk 5]
```diff
@@ -2329,6 +2329,21 @@ def testSymmetricProduct(self, shape, dtype, symmetrize_output):
...
+    kind=["symmetric", "lower", "upper"],
+  )
+  @jax.default_matmul_precision("float32")
+  def testPascal(self, n, kind):
+    args_maker = lambda: []
+    osp_fun = partial(osp.linalg.pascal, n=n, kind=kind, exact=False)
+    jsp_fun = partial(jsp.linalg.pascal, n=n, kind=kind)
+    self._CheckAgainstNumpy(osp_fun,
+                            jsp_fun, args_maker,
+                            atol=1e-3,
```
[the remark this answers is not in the document]
Comment (mattbahr) [reply tests/linalg_test.py:2343]: Good idea. Setting `rtol=1e-3` also allowed me to test with `n=20`, which was previously failing on my system. I went ahead and added that test case so it covers the same dimensions as the Hilbert test. If it causes problems I can remove.

**Review by jakevdp — APPROVED** [review]

**Comment by jakevdp** [comment]
Sorry about this, but now we're seeing failures on TPU! We'll need `rtol=0.01` or so when running on TPU! You can do device-specific tolerances using this pattern: https://github.com/jax-ml/jax/blob/ed75189c921a66e1d5232923ff5b5cdbc23e766f/tests/linalg_test.py#L742

**Comment by mattbahr** [comment]
@jakevdp No problem! Learning a lot! Just updated

**Comment by mattbahr** [comment]
Hi @jakevdp just a friendly ping to hopefully keep this one from getting lost in the shuffle :)

**Review by jakevdp — APPROVED** [review]
